File: src/main/java/dev/sky_lock/pocketlifevehicle/util/TypeChecks.java
package dev.sky_lock.pocketlifevehicle.util;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

/**
 * @author sky_lock
 */

public class TypeChecks {

    public static <E> List<E> checkListTypeDynamically(Object listObj, Class<E> type) {
        return Optional.of(listObj)
                .filter(List.class::isInstance)
                .map(List.class::cast)
                .map(list -> checkTypeDynamically(list, type))
                .orElseThrow(ClassCastException::new);
    }

    private static <E> List<E> checkTypeDynamically(List<?> uncheckedList, Class<E> type) {
        return uncheckedList.stream()
                .filter(type::isInstance)
                .map(type::cast)
                .collect(Collectors.toList());
    }

}

File: src/main/kotlin/dev/sky_lock/pocketlifevehicle/util/TypeChecks.kt
package dev.sky_lock.pocketlifevehicle.util

import java.util.*
import java.util.stream.Collectors

/**
 * @author sky_lock
 */
object TypeChecks {
    fun <E> checkListTypeDynamically(listObj: Any, type: Class<E>): List<E> {
        return Optional.of(listObj)
                .filter { obj: Any? -> MutableList::class.java.isInstance(obj) }
                .map<List<*>> { obj: Any? -> MutableList::class.java.cast(obj) }
                .map { list: List<*> -> checkTypeDynamically(list, type) }
                .orElseThrow { ClassCastException() }
    }

    private fun <E> checkTypeDynamically(uncheckedList: List<*>, type: Class<E>): List<E> {
        return uncheckedList.stream()
                .filter { obj: Any? -> type.isInstance(obj) }
                .map { obj: Any? -> type.cast(obj) }
                .collect(Collectors.toList())
    }
}